File: crates/mithril-oxide-sys/src/Dialect/LLVMIR/LLVMTypes.rs
pub use self::ffi::{
    LLVMArrayType_get, LLVMFixedVectorType_get, LLVMFunctionType_get, LLVMMetadataType_get,
    LLVMPPCFP128Type_get, LLVMPointerType_get, LLVMScalableVectorType_get,
    LLVMStructType_getLiteral, LLVMTokenType_get, LLVMVoidType_get, LLVMX86MMXType_get,
};

#[cxx::bridge]
mod ffi {
    #[namespace = "mlir"]
    unsafe extern "C++" {
        include!("mithril-oxide-sys/cpp/Dialect/LLVMIR/LLVMTypes.hpp");

        type MLIRContext = crate::IR::MLIRContext::MLIRContext;
    }

    #[namespace = "mithril_oxide_sys::llvm"]
    unsafe extern "C++" {
        include!("mithril-oxide-sys/cpp/Dialect/LLVMIR/LLVMTypes.hpp");

        type c_void = crate::IR::Value::ffi::c_void;

        pub unsafe fn LLVMArrayType_get(
            elementType: *const c_void,
            numElements: u32,
        ) -> *const void;
        pub unsafe fn LLVMFixedVectorType_get(
            elementType: *const c_void,
            numElements: u32,
        ) -> *const void;
        pub unsafe fn LLVMFunctionType_get(
            result: *const c_void,
            arguments: &[*const c_void],
            isVarArg: bool,
        ) -> *const void;
        pub unsafe fn LLVMMetadataType_get(ctx: *mut MLIRContext) -> *const void;
        pub unsafe fn LLVMPointerType_get(ctx: *const c_void) -> *const void;
        pub unsafe fn LLVMPPCFP128Type_get(ctx: *mut MLIRContext) -> *const void;
        pub unsafe fn LLVMScalableVectorType_get(
            elementType: *const c_void,
            numElements: u32,
        ) -> *const void;
        pub unsafe fn LLVMStructType_getLiteral(
            ctx: *mut MLIRContext,
            fields: &[*const c_void],
            isPacked: bool,
        ) -> *const void;
        pub unsafe fn LLVMTokenType_get(ctx: *mut MLIRContext) -> *const void;
        pub unsafe fn LLVMVoidType_get(ctx: *mut MLIRContext) -> *const void;
        pub unsafe fn LLVMX86MMXType_get(ctx: *mut MLIRContext) -> *const void;
    }
}

File: crates/mithril-oxide-sys/cpp/Dialect/LLVMIR/LLVMTypes.cpp
#include "LLVMTypes.hpp"

#include <mlir/Dialect/LLVMIR/LLVMTypes.h>
#include <mlir/IR/Types.h>
#include <mlir/Support/LLVM.h>
#include <rust/cxx.h>


namespace mithril_oxide_sys::llvm {

const void *LLVMArrayType_get(const void *elementType, unsigned int numElements)
{
    return LLVMArrayType::get(
        mlir::Type::getFromOpaquePointer(elementType),
        numElements
    ).getAsOpaquePointer();
}

const void *LLVMFixedVectorType_get(const void *elementType, unsigned int numElements)
{
    return LLVMFixedVectorType::get(
        mlir::Type::getFromOpaquePointer(elementType),
        numElements
    ).getAsOpaquePointer();
}

const void *LLVMFunctionType_get(const void *result, rust::Slice<const void*> arguments, bool isVarArg)
{
    return LLVMFunctionType::get(
        mlir::Type::getFromOpaquePointer(result),
        mlir::ArrayRef(reinterpret_cast<mlir::Type *>(arguments.data()), arguments.length()),
        isVarArg
    ).getAsOpaquePointer();
}

const void *LLVMMetadataType_get(MLIRContext *ctx)
{
    return LLVMMetadataType::get(ctx).getAsOpaquePointer();
}

const void *LLVMPointerType_get(
    MLIRContext *ctx,
    const void *elementType,
    unsigned int addressSpace
)
{
    if (elementType == nullptr) {
        return LLVMPointerType::get(ctx, addressSpace).getAsOpaquePointer();
    } else {
        return LLVMPointerType::get(
            ctx,
            mlir::Type::getFromOpaquePointer(elementType),
            addressSpace
        ).getAsOpaquePointer();
    }
}

const void *LLVMPPCFP128Type_get(MLIRContext *ctx)
{
    return LLVMPPCFP128Type::get(ctx).getAsOpaquePointer();
}

const void *LLVMScalableVectorType_get(const void *elementType, unsigned int numElements)
{
    return LLVMScalableVectorType::get(
        mlir::Type::getFromOpaquePointer(elementType),
        numElements
    ).getAsOpaquePointer();
}

const void *LLVMStructType_getLiteral(
    MLIRContext *ctx,
    rust::Slice<const void*> fields,
    bool isPacked
)
{
    return LLVMStructType::getLiteral(
        ctx,
        mlir::ArrayRef(reinterpret_cast<mlir::Type *>(fields.data()), fields.length()),
        isPacked
    ).getAsOpaquePointer();
}

const void *LLVMTokenType_get(MLIRContext *ctx)
{
    return LLVMTokenType::get(ctx).getAsOpaquePointer();
}

const void *LLVMVoidType_get(MLIRContext *ctx)
{
    return LLVMVoidType::get(ctx).getAsOpaquePointer();
}

const void *LLVMX86MMXType_get(MLIRContext *ctx)
{
    return LLVMX86MMXType::get(ctx).getAsOpaquePointer();
}

} // namespace mithril_oxide_sys::llvm
